### PR TITLE
add sleep_control playbook

### PR DESCRIPTION
**Pull request by kbroughton**
This playbook is intended to reduce costs and save energy by allowing mantl clusters to be put to sleep indefinitely or on a cron cycle.
It's useful for devs working on mantl who don't want to destroy a WIP cluster but won't be touching it while they sleep or over the weekend.
A stack with a few m4.xlarge's could cost 2k per month, so one could be saving upwards of $700 per month.  Although you still pay for eips, block storage etc, typically the running instance cost is 90% of the bill.
It also tests resilience of the cluster regularly and forces users to run stateless containers backed by proper distributed stores like cassandra or mongo.

This requires elastic ips, so it is dependent on #1283  "replace public ips with elastic ips so they remain fixed on reboots"

### CODE REVIEW

**Comment by kbroughton** [comment]
After several successful restarts the consul cluster ended up in a bad state with no leader.  Ansible fails in consul role waiting for leader.

Looking at the [recovery page for consul](https://www.consul.io/docs/guides/outage.html) it appears this PR will need more work.
Keep in mind that the rebuilt server needs to have the same IP as the failed server. Again, once this server is online, the cluster will return to a fully healthy state.

Two options, we support both:
1. Just don't power down the consul servers.  There are only three and typically will by on small-cheap machines.  This is the default.
2.  systemctl stop consul; replace the raft/peers.json with 

Finally, we have to run mantlui role again as those containers don't auto-restart.
I added a debug message for the same, as we can't include roles with handlers until 
 https://github.com/CiscoCloud/mantl/issues/1343#issuecomment-209117859
is resolved

**Comment by siddharthist** [comment]
@kbroughton You might want to check out https://github.com/CiscoCloud/mantl/pull/948, it is related to the whole Consul reset process.

**Comment by kbroughton** [comment]
Ah, very nice!  The abp playbook is essentially what I have in my sleep_control startup block.
If this PR gets added, i would replace my block with a call to his.

One thing I had to do was modify the peers.json content with a default(groups['all']) for when consul_dc_group is not set.  

{% for host in groups[consul_servers_group] | intersect(groups[consul_dc_group] | default(groups['all'])) %}"{{ hostvars[host].private_ipv4 }}:8300"{% if not loop.last %}, 
{% endif %}{% endfor %}

You would essentially just get abps functionality if you run mine with -e do_sleep=no.

I've probably got more serialization than is necessary, but I've been able to successfully powered down my mantl cluster several times now.  Next is to add a chronos job to fire it off automatically every night if there are no recent ssh sessions or api calls.

**Comment by kbroughton** [comment]
Just a comment here.  This playbook is solving all kinds of problems for me.
For example, I recently rotated certs to add more wildcards using my PR https://github.com/CiscoCloud/mantl/pull/1307

Consul was totally broken afterwards.

I ran 
`
rm -rm ssl
git checkout ssl
./security-setup --dns_alt_cert_names_string "DNS:*.extra.dns1, DNS:*.extra.dns2"
ansible-playbook -e @security.yml mantl.yml

This runs as far as consul handlers and then fails to find a leader, new certs are copied over.
Then I run sleep control to get the consul cluster health back.

ansible-playbook -e @security.yml  playbooks/aws/sleep_control.yml -e do_sleep=no -e override_tags=yes -e do_restart_consul=yes

and then i run mant.yml again which now goes to completion.   Voila.  cert rotation.
`

**Comment by siddharthist** [comment]
@kbroughton Is there a way to make this playbook not rely on elastic ips?

**Comment by kbroughton** [comment]
I don't believe it relies on elastic ips at the moment.  I just recommend not rebooting machines if the ips are public (not elastic) since they might not come back the same, and this script does not validate dns/route53 is correct.

I have been using it with terraform.py using the modified branch supporting public ips.  And using a private only deployment with loadbalancing following https://github.com/CiscoCloud/mantl/pull/1421 with a vpn connection to the private ips.

For the latter, we are running on a 3rd party hosted aws vpc and it is VERY tricky getting terraform to work with an existing deployment, so I modified sleep_control.yml to sleep_control_static.yml, included in the last push, to support static ansible inventory.  This seems to work well.  Just having some issues with zookeeper authentication, possibly due to rotating security.yml.  I would highly recommend going the all-private with load balancing route.

The two scripts should be merged and cleaned up a lot, but it is a bit tricky making ansible/terraform work with non-running hosts and finding ansible modules that work without ssh-to-ip access.
Holding back because I think it would be best to modify either terraform.py or ec2.py so that they name variables the same (possibly just adding duplicate alias entries to either or both).  For example, terraform.py produces private_ipv4, but that is not available with ec2.py so i have to set a fact, but then I can't use --start-at-task.

**Comment by siddharthist** [comment]
@kbroughton We definitely don't want to include any GPLv3 source code in this repo, could that file be downloaded instead?

File: plugins/inventory/terraform.py [hunk 1]
```diff
@@ -399,6 +399,7 @@ def aws_host(resource, module_name):
         # generic
         'public_ipv4': raw_attrs['public_ip'],
         'private_ipv4': raw_attrs['private_ip'],
+        'instance_state': raw_attrs['instance_state'],
```
Comment (siddharthist) [inline plugins/inventory/terraform.py]: Changes to terraform.py should go here: https://github.com/CiscoCloud/terraform.py

File: plugins/inventory/terraform_hosts [hunk 2]
```diff
@@ -0,0 +1,4 @@
+[control]
```
Comment (siddharthist) [inline plugins/inventory/terraform_hosts]: What is this file used for?

File: playbooks/aws/sleep_control.yml [hunk 3]
```diff
@@ -0,0 +1,326 @@
...
+  - name: stop consul 
+    shell: systemctl stop consul
+    become: yes
+    delegate_to: "{{ item }}"
+    with_items: "{{ groups['role=control'] }}"
+    when: ( do_restart_consul == 'yes' ) or ( do_sleep_controllers == 'yes' )
+
+  - name: template raft/peers.json to replace null from graceful shutdown 
+    template: 
+      src: ../../roles/consul/templates/peers.json.j2
```
Comment (siddharthist) [inline playbooks/aws/sleep_control.yml]: This sort of thing makes me think that this playbook should have its own role.

File: library/ec2_remote_facts.py [hunk 4]
```diff
@@ -0,0 +1,170 @@
+#!/usr/bin/python
+#
+# This is a free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
```
Comment (siddharthist) [inline library/ec2_remote_facts.py]: We definitely don't want to include any GPLv3 source code in this repo, could this file be downloaded instead?

**Comment by siddharthist** [comment]
Closing since this doesn't line up with out defaults (no elastic IPs) and is very complex.

**Pull request closed without merging**